### PR TITLE
feat: add robust validation and 100% branch coverage for _derive_grade

### LINKED ISSUE #10: QA-01: Unit Test Coverage for Grade Derivation Logic
### Task Description

The _derive_grade utility in main.py handles critical business logic. We need to ensure 100% branch coverage to prevent regressions in scoring.

### Context and Background

Target: main._derive_grade
Location: tests/test_ci.py
Test Suite: Use pytest with @pytest.mark.parametrize for edge cases.

### Acceptance Criteria

Boundary values (e.g., exactly 90% for an 'A').
Type handling (ensure ValueError is raised for non-numeric inputs).
Negative/Over-scale inputs (input > 100 or < 0).

### Hints

_No response_

### PR DESCRIPTION
**Pull request by BHOGALA-SRIKA**
Closes #10

## Description

This PR resolves the issue tracking critical business logic testing for the backend grading utility. It introduces comprehensive validation checking and branch coverage for `_derive_grade` in `main.py` to prevent regression bugs.

### Changes Made
- **Type Validation:** Added explicit type checks using `isinstance` to catch and gracefully reject non-numeric inputs (`str`, `NoneType`, `list`, `bool`), raising a clean `ValueError` instead of unhandled runtime crashes.
- **Scale Boundaries:** Implemented strict value limits ensuring inputs reside safely between `0` and `100` inclusive.
- **Test Isolation Mocks:** Configured `sys.modules` interception in `tests/test_ci.py` to isolate tests from heavy third-party framework layers (`torch` binary paths and `supabase` network connections), dropping local CI suite test execution to < 2 seconds.

## Checklist

- [ ] `npm run lint` passes with no errors
- [ ] `npm run build` compiles without TypeScript errors
- [x] `python -m pytest` passes (including new tests I added)
- [x] No `.env` files, API keys, secrets, model weights, or `__pycache__` in this diff
- [x] Branch is rebased on `main`, not merged

### CODE REVIEW

File: backend/tests/test_ci.py [hunk 1]
```diff
@@ -100,3 +115,80 @@ def test_dev_bypass_constants_are_readable():
...
+sys.modules['supabase'] = mock_supabase
+
+# 2. Trick Python into bypassing 'inference' entirely to avoid the broken PyTorch DLLs
+mock_inference = MagicMock()
+# Provide dummy mock functions so main.py can unpack them during import
+mock_inference.load_models = MagicMock()
+mock_inference.predict_stream_a = MagicMock()
+mock_inference.predict_stream_b = MagicMock()
+sys.modules['inference'] = mock_inference
+
```
Comment (jpdevhub) [inline backend/tests/test_ci.py]: This code is dublicated since you already correctly placed this mock interception at the very top of the file (lines 14-27) where it belongs, the block at the bottom is redundant and should be deleted.

**Review by jpdevhub — COMMENTED** [review]
Excellent work on this PR! The strict type validation for the grading boundaries is a great addition, and the way you isolated the heavy ML modules using sys.modules interception in the tests is fantastic. I confirmed it drops the test suite execution time to less than a second.
Once that duplicate block at the bottom of the test file is removed, this is good to go!

File: backend/tests/test_ci.py (hunk 1 above)
Comment (BHOGALA-SRIKA) [reply backend/tests/test_ci.py]: Thanks for the review. I caught that duplicate block, cleaned it up, and just pushed the final refactored file structure. It should be 100% clean and ready now.

**Review by jpdevhub — APPROVED** [review]
Thanks for contributing.
